cmd/tiller-proxy/main.go

```diff
@@ -180,9 +183,32 @@ func main() {
 		port = "8080"
 	}
 	addr := ":" + port
-	log.WithFields(log.Fields{"addr": addr}).Info("Started Tiller Proxy")
-	err = http.ListenAndServe(addr, n)
-	if err != nil {
-		log.Fatalf("Unable to start the server: %v", err)
+
+	srv := &http.Server{
+		Addr:    addr,
+		Handler: n,
 	}
+
+	go func() {
+		log.WithFields(log.Fields{"addr": addr}).Info("Started Tiller Proxy")
+		err = srv.ListenAndServe()
+		if err != nil {
+			log.Info(err)
+		}
+	}()
+
+	// Catch SIGINT and SIGKILL
+	// Set up channel on which to send signal notifications.
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+	log.Info("Set system to get notified on signals")
+	s := <-c
+	log.Infof("Received signal: %v. Waiting for existing requests to finish", s)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*60)
```

Comment:
hmm, if I understand this correctly, we have a built in termination deadline that will wait 60 seconds for things to gracefully exit, and after the timeout will force shutdown?

The way terminationGracePeriodSeconds and `docker stop` work is that they will send a SIGTERM followed by a SIGKILL after the specified time. Given this, we don't actually need to build in our own timer (though it doesn't necessarily hurt). However, it might be better to simply wait indefinitely (if even possible, or use a much longer timeout if not) for requests to finish, and rely on Kubernetes to send us the SIGKILL after the configured time. The reason for this is that we now will not support it if terminationGracePeriodSeconds > 60s. I don't think it's a big deal, but worth considering.